crossbeam/property_signatures/property_signatures.py

```diff
@@ -0,0 +1,474 @@
+         for i in inputs), [])
+  else:
+    if len(inputs) > FIXED_NUM_INPUTS:
+      inputs = inputs[:FIXED_NUM_INPUTS]
+    result = _basic_signature(output, fixed_length)
+    basic_sig_length = len(result)
+    result.extend(sum(
+        (_basic_signature(i, fixed_length) + _compare(i, output, fixed_length)
+         for i in inputs), []))
+    if len(inputs) < FIXED_NUM_INPUTS:
+      assert (len(result) - basic_sig_length) % len(inputs) == 0
+      result.extend([None] * (
+          (len(result) - basic_sig_length) // len(inputs) *  # Length per input
+          (FIXED_NUM_INPUTS - len(inputs))))  # Number of inputs to pad
+    return  result
+
+
+def _property_signature_single_object(
+    x: Any,
+    output: Any,
+    fixed_length: bool = True) -> List[Optional[bool]]:
+  """Returns a property signature for an object in context of an I/O example."""
+  return _basic_signature(x, fixed_length) + _compare(x, output, fixed_length)
+
+
+def _reduce_across_examples(
+    signatures: List[List[Optional[bool]]]
+) -> List[Tuple[float, bool, bool, float]]:
+  """Reduce across examples (frac applicable, all True?, all False?, frac True)."""
+  # All signatures should have the same length across examples.
+  num_examples = len(signatures)
+  assert num_examples
+  signature_len = len(signatures[0])
+  assert all(len(sig) == signature_len for sig in signatures)
+
+  result = []
+  for bools in zip(*signatures):
+    bools_not_none = [x for x in bools if x is not None]
+    num_not_none = len(bools_not_none)
+    frac_true = sum(bools_not_none) / num_not_none if num_not_none else 0.5
+    result.append((len(bools_not_none) / num_examples,
+                   frac_true == 1,
+                   frac_true == 0,
+                   frac_true))
+  return result
+
+
+def property_signature_io_examples(
+    input_values: List[value_module.Value],
+    output_value: value_module.Value,
+    fixed_length: bool = True) -> List[Tuple[float, bool, bool, float]]:
+  """Returns a property signature for a set of I/O examples."""
+  num_examples = output_value.num_examples
+  assert all(i_value.num_examples == num_examples for i_value in input_values)
+  return _reduce_across_examples(
+      [_property_signature_single_example(  # pylint: disable=g-complex-comprehension
+          [i_value[example_index] for i_value in input_values],
+          output_value[example_index], fixed_length=fixed_length)
+       for example_index in range(num_examples)])
+
+IO_EXAMPLES_SIGNATURE_LENGTH = len(property_signature_io_examples(
+    [value_module.InputVariable([1, 2], 'in1')],
+    value_module.OutputValue([-1, -2]),
+    fixed_length=True))
+
+
+def _property_signature_concrete_value(
+    value: value_module.Value,
+    output_value: value_module.Value,
+    fixed_length: bool = True) -> List[Tuple[float, bool, bool, float]]:
+  """Returns a property signature for a value w.r.t. a set of I/O examples."""
+  assert value.num_free_variables == 0
```

Comment:
minor: also need `assert value.num_examples==output_value.num_examples`

Reply:
actually this isn't always true. constants (and expressions that are effectively constant) have num_examples==1 and we "broadcast" the single example to fit the multiple examples when combining/comparing to other values with higher num_examples

Reply:
got it!